src/interface/swarm.cpp

```diff
@@ -0,0 +1,347 @@
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <algorithm>
+#include <cstdlib>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "bvals/cc/bvals_cc.hpp"
+#include "globals.hpp" // my_rank
+#include "mesh/mesh.hpp"
+#include "swarm.hpp"
+
+namespace parthenon {
+
+Swarm::Swarm(const std::string label, const Metadata &metadata, const int nmax_pool_in)
+    : label_(label), m_(metadata), nmax_pool_(nmax_pool_in),
+      mask_("mask", nmax_pool_, Metadata({Metadata::Boolean})),
+      marked_for_removal_("mfr", nmax_pool_, Metadata({Metadata::Boolean})),
+      mpiStatus(true) {
+  Add("x", Metadata({Metadata::Real}));
+  Add("y", Metadata({Metadata::Real}));
+  Add("z", Metadata({Metadata::Real}));
+  num_active_ = 0;
+  max_active_index_ = 0;
+
+  auto mask_h = mask_.data.GetHostMirror();
+  auto marked_for_removal_h = marked_for_removal_.data.GetHostMirror();
+
+  for (int n = 0; n < nmax_pool_; n++) {
+    mask_h(n) = false;
+    marked_for_removal_h(n) = false;
+    free_indices_.push_back(n);
+  }
+
+  mask_.data.DeepCopy(mask_h);
+  marked_for_removal_.data.DeepCopy(marked_for_removal_h);
+}
+
+void Swarm::Add(const std::vector<std::string> labelArray, const Metadata &metadata) {
+  // generate the vector and call Add
+  for (auto label : labelArray) {
+    Add(label, metadata);
+  }
+}
+
+std::shared_ptr<Swarm> Swarm::AllocateCopy(const bool allocComms, MeshBlock *pmb) {
+  Metadata m = m_;
+
+  auto swarm = std::make_shared<Swarm>(label(), m, nmax_pool_);
+
+  return swarm;
+}
+
+///
+/// The internal routine for allocating a particle swarm.  This subroutine
+/// is topology aware and will allocate accordingly.
+///
+/// @param label the name of the variable
+/// @param metadata the metadata associated with the particle
+void Swarm::Add(const std::string label, const Metadata &metadata) {
+  // labels must be unique, even between different types of data
+  if (intMap_.count(label) > 0 || realMap_.count(label) > 0) {
+    throw std::invalid_argument("swarm variable " + label +
+                                " already enrolled during Add()!");
+  }
+
+  if (metadata.Type() == Metadata::Integer) {
+    auto var = std::make_shared<ParticleVariable<int>>(label, nmax_pool_, metadata);
+    intVector_.push_back(var);
+    intMap_[label] = var;
+  } else if (metadata.Type() == Metadata::Real) {
+    auto var = std::make_shared<ParticleVariable<Real>>(label, nmax_pool_, metadata);
+    realVector_.push_back(var);
+    realMap_[label] = var;
+  } else {
+    throw std::invalid_argument("swarm variable " + label +
+                                " does not have a valid type during Add()");
+  }
+}
+
+void Swarm::Remove(const std::string label) {
+  int idx, isize;
+  bool found = false;
+
+  // Find index of variable
+  idx = 0;
```

Comment:
Get rid of idx above and simply use this line:

`int idx = 0;` 

That will reduce the number of lines of code.